tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/BlockingMiddleware.cs

```diff
@@ -0,0 +1,116 @@
+    {
+        _next = next;
+        _runSecurity = runSecurity;
+        _endPipeline = endPipeline;
+    }
+
+    private static Task WriteResponse(HttpContext context, SecuritySettings settings, out bool endedResponse)
+    {
+        var httpResponse = context.Response;
+        if (!httpResponse.HasStarted)
+        {
+            httpResponse.Clear();
+            foreach (var cookie in context.Request.Cookies)
+            {
+                httpResponse.Cookies.Delete(cookie.Key);
+            }
+
+            httpResponse.Headers.Clear();
+            httpResponse.StatusCode = 403;
+            var template = settings.BlockedJsonTemplate;
+            if (context.Request.Headers["Accept"].ToString().Contains("text/html"))
+            {
+                httpResponse.ContentType = "text/html";
+                template = settings.BlockedHtmlTemplate;
+            }
+            else
+            {
+                httpResponse.ContentType = "application/json";
+            }
+
+            endedResponse = true;
+            return httpResponse.WriteAsync(template);
+        }
+
+        context.Response.Body.Dispose();
+        endedResponse = true;
+        return Task.CompletedTask;
+    }
+
+    internal async Task Invoke(HttpContext context)
+    {
+        var security = Security.Instance;
+        var endedResponse = false;
+        if (security.Settings.Enabled)
+        {
+            var securityTransport = new SecurityTransport(security, context, Tracer.Instance.ActiveScope.Span as Span);
+            if (_runSecurity)
+            {
+                if (_endPipeline && !context.Response.HasStarted)
+                {
+                    context.Response.StatusCode = 404;
+                }
+
+                using var result = securityTransport.ShouldBlock();
+                if (result.ShouldBeReported)
+                {
+                    if (result.Block)
+                    {
+                        await WriteResponse(context, security.Settings, out endedResponse).ConfigureAwait(false);
+                        securityTransport.MarkBlocked();
+                    }
+
+                    securityTransport.Report(result, endedResponse);
+                }
+            }
+
+            if (_next != null && !endedResponse)
+            {
+                // unlikely that security is disabled and there's a blockexception, but might happen as race condition
+                try
+                {
+                    await _next(context).ConfigureAwait(false);
+                }
+                catch (BlockException e)
+                {
+                    await WriteResponse(context, security.Settings, out endedResponse).ConfigureAwait(false);
+                    if (!e.Reported)
+                    {
+                        securityTransport.Report(e.Result, endedResponse);
+                    }
+
+                    securityTransport.Cleanup();
+                }
+            }
+        }
```

Comment:
If security it disabled, we won't be running here would we? Or is this handling the case where RCM disables security once we're already in here? If that's the case, I think the `else` block also needs to catch a `BlockException`?

Reply:
good point!! added the `Security.Enabled` check here too in 60213612bd9dc48fe4f254b95d896f2a95e759d4